mapreduce/src/main/java/uk/gov/gchq/palisade/PalisadeRecordReader.java

```diff
@@ -0,0 +1,229 @@
+        }
+        resourceDetails = reqDetails;
+        resIt = reqDetails.getResources().entrySet().iterator();
+        serialiser = PalisadeInputFormat.getSerialiser(taskAttemptContext);
+        currentKey = null;
+        currentValue = null;
+        processed = 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean nextKeyValue() throws IOException {
+        //if we don't have a current item iterator or it's empty...
+        while (itemIt == null || !itemIt.hasNext()) {
+            //...try to move to the next one, if we can't then we're done
+            if (!moveToNextResource()) {
+                return false;
+            }
+        }
+        //by now we've either got an item iterator with results or returned
+        currentValue = itemIt.next();
+        return true;
+    }
+
+    /**
+     * Move to the next resource in the list of resources for this input split. This co-ordinates the logic for moving
+     * to the next resource.
+     *
+     * @return true iff we have a successfully moved to a new data stream for the next resource
+     */
+    private boolean moveToNextResource() {
+        //do we have a resource iterator?
+        if (resIt == null) {
+            return false;
+        } else {
+            //any resources left to process?
+            if (resIt.hasNext()) {
+                //set up the next resource
+                setupItemStream();
+                return true;
+            } else {
+                //end of things to be iterated
+                resIt = null;
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Internal method to move to the next resource in our iterator of resources. This makes the actual call to the data
+     * service and waits for the request to complete before extracting the data stream iterator which clients will use.
+     * If successful there will be a new item iterator set up and ready to retrieve data.
+     */
+    private void setupItemStream() {
+        Map.Entry<Resource, ConnectionDetail> entry = resIt.next();
+        final Resource resource = entry.getKey();
+        final ConnectionDetail conDetails = entry.getValue();
+        //stash the resource
+        currentKey = resource;
+        final DataService service = conDetails.createService();
+        //create the singleton resource request for this resource
+        final DataRequestResponse singleResourceRequest = new DataRequestResponse(resourceDetails.getRequestId(),
+                resource, conDetails);
+        //lodge request with the data service
+        final CompletableFuture<ReadResponse<Object>> futureResponse = service.read(new ReadRequest(singleResourceRequest));
+        //when this future completes, we should have an iterator of things once we deserialise
+        itemIt = futureResponse.thenApply(response -> response.getData().map(serialiser::deserialise).iterator()).join();
```

Comment:
Perhaps separate the .join() call onto a separate line and explain that the code will block at this point until the response has been and deserialised.

Reply:
Done